File: internal/session/resume_test.go
package session

import (
	"reflect"
	"testing"
)

func TestResumeArgvKnowsTheTwoAgentsItHasBeenMeasuredAgainst(t *testing.T) {
	for _, tc := range []struct {
		name   string
		launch []string
		want   []string
	}{
		{"claude", []string{"claude"}, []string{"claude", "--continue"}},
		{
			"claude keeps the arguments it was given",
			[]string{"claude", "--model", "opus"},
			[]string{"claude", "--continue", "--model", "opus"},
		},
		{
			"claude by absolute path is still claude",
			[]string{"/home/x/.local/bin/claude"},
			[]string{"/home/x/.local/bin/claude", "--continue"},
		},
		// A subcommand, so it goes first. `codex --model x resume --last`
		// is not a command line codex accepts.
		{"codex", []string{"codex"}, []string{"codex", "resume", "--last"}},
		{
			"codex keeps its arguments, after the subcommand",
			[]string{"codex", "--model", "gpt"},
			[]string{"codex", "resume", "--last", "--model", "gpt"},
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			got, ok := ResumeArgv(tc.launch)
			if !ok {
				t.Fatalf("ResumeArgv(%v) declined; this is an agent that resumes", tc.launch)
			}
			if !reflect.DeepEqual(got, tc.want) {
				t.Errorf("ResumeArgv(%v) = %v, want %v", tc.launch, got, tc.want)
			}
		})
	}
}

// The input argv must come back untouched.
//
// concat exists for this: appending onto launch[:1] shares the caller's backing
// array, so building the resumed argv would overwrite the caller's launch[1] --
// and the caller here is a restore that goes on to show the user what it ran.
func TestResumeArgvDoesNotEditWhatItWasGiven(t *testing.T) {
	launch := []string{"claude", "--model", "opus"}
	if _, ok := ResumeArgv(launch); !ok {
		t.Fatal("ResumeArgv declined claude")
	}
	if !reflect.DeepEqual(launch, []string{"claude", "--model", "opus"}) {
		t.Errorf("the caller's argv was rewritten in place: %v", launch)
	}
}

func TestResumeArgvDeclinesWhatItCannotBringBack(t *testing.T) {
	for _, tc := range []struct {
		name   string
		launch []string
	}{
		// A row from before the panel recorded argvs. It comes back as a login
		// shell, and there is nothing to resume about a login shell.
		{"nothing recorded", nil},
		{"a login shell", []string{"bash", "-l"}},
		{"an ordinary command", []string{"npm", "run", "dev"}},
		// Launched asking for it already. Two --continue flags is at best
		// noise on the command line the dialog prints.
		{"claude already continuing", []string{"claude", "--continue"}},
		{"claude already continuing, short", []string{"claude", "-c"}},
		{"claude resuming a named session", []string{"claude", "--resume", "abc"}},
		{"codex already resuming", []string{"codex", "resume", "--last"}},
		// Not measured against, so not guessed at. The cost of being wrong is
		// a pane that dies at startup.
		{"opencode", []string{"opencode"}},
		// The program is what decides, not a substring of an argument:
		// `--model claude-opus-5` is not claude.
		{"an argument that merely mentions an agent", []string{"env", "X=claude", "sh"}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if got, ok := ResumeArgv(tc.launch); ok {
				t.Errorf("ResumeArgv(%v) = %v, want a refusal", tc.launch, got)
			}
			if ResumesConversation(tc.launch) {
				t.Errorf("ResumesConversation(%v) said yes", tc.launch)
			}
			if p := ResumeProgram(tc.launch); p != "" {
				t.Errorf("ResumeProgram(%v) = %q, want empty", tc.launch, p)
			}
		})
	}
}

// Two agents restored into one directory would resume the same conversation.
//
// Both flags mean "the most recent conversation here", so this is not two
// conversations coming back; it is one conversation coming back twice, with
// nothing on either screen to say so.
func TestTwoSessionsInOneDirectoryCannotBothResume(t *testing.T) {
	rows := []ResumeCandidate{
		{Launch: []string{"claude"}, CWD: "/w/api"},
		{Launch: []string{"claude"}, CWD: "/w/api"},
		{Launch: []string{"claude"}, CWD: "/w/web"},
		{Launch: []string{"codex"}, CWD: "/w/api"},
		{Launch: []string{"bash", "-l"}, CWD: "/w/api"},
	}

	if !ResumeIsAmbiguous(rows, "claude", "/w/api") {
		t.Error("two claude sessions in /w/api were not reported as ambiguous")
	}
	// One of a kind in its directory: nothing to be confused with.
	if ResumeIsAmbiguous(rows, "claude", "/w/web") {
		t.Error("the only claude session in /w/web was reported as ambiguous")
	}
	// A different agent in the same directory is a different transcript.
	if ResumeIsAmbiguous(rows, "codex", "/w/api") {
		t.Error("the only codex session in /w/api was reported as ambiguous")
	}
	// A shell in the directory is not a claimant on anybody's conversation.
	if ResumeIsAmbiguous(rows, "bash", "/w/api") {
		t.Error("a login shell was treated as something that resumes")
	}
}

// A row with no directory recorded cannot be reasoned about, and answering
// "ambiguous" for it would turn every such row into a cold start on the
// strength of a missing field.
func TestAmbiguityNeedsBothHalvesOfTheKey(t *testing.T) {
	rows := []ResumeCandidate{
		{Launch: []string{"claude"}, CWD: ""},
		{Launch: []string{"claude"}, CWD: ""},
	}
	if ResumeIsAmbiguous(rows, "claude", "") {
		t.Error("an empty directory matched itself")
	}
	if ResumeIsAmbiguous(rows, "", "/w/api") {
		t.Error("an empty program matched itself")
	}
}

File: web/src/components/resume.test.ts
import { readFileSync } from 'node:fs'
import { describe, expect, it } from 'vitest'

import { resumeArgv, resumeIsAmbiguous, resumeProgram, restoreLaunch } from './resume'

/**
 * The rule is written twice, in two languages, and this is what stops the two
 * from drifting.
 *
 * The server decides what a restored session runs and the dialog prints it
 * before anybody presses the button. If the browser's table says `claude` and
 * the server's says `codex`, or one of them learns a third agent, the dialog
 * spends its whole existence lying about the one thing it is for — and nothing
 * would say a word, because both sides type-check and both sides run.
 *
 * Crude on purpose, in the manner of dirpicker.test.ts: it reads the Go file as
 * text. A crude check that runs on every commit is worth more than an exact one
 * that needs a Go toolchain in vitest.
 */
describe('the table on the other side of the wire', () => {
  const go = readFileSync(new URL('../../../internal/session/resume.go', import.meta.url), 'utf8')

  it('finds the Go file at all', () => {
    // Without this the assertions below pass by searching an empty string,
    // which is the failure mode of every checker that reads a file by path.
    expect(go).toContain('func ResumeArgv')
  })

  it('resumes exactly the agents Go resumes', () => {
    // The `case "x":` arms of ResumeArgv's switch, which is the list.
    const arms = [...go.matchAll(/^\tcase "([a-z-]+)":$/gm)].map((m) => m[1])
    expect(arms.length).toBeGreaterThan(0)
    for (const agent of arms) {
      expect(
        resumeArgv([agent]),
        `${agent} resumes in Go and not in the browser`,
      ).not.toBeNull()
    }
    // And nothing this side has invented on its own. opencode is the one that
    // matters: the panel launches it, so it is the obvious thing to add here
    // without measuring it.
    for (const agent of ['opencode', 'aider', 'gemini']) {
      if (arms.includes(agent)) continue
      expect(resumeArgv([agent]), `${agent} resumes in the browser and not in Go`).toBeNull()
    }
  })

  it('spells the flags the same way Go does', () => {
    expect(resumeArgv(['claude'])).toEqual(['claude', '--continue'])
    expect(go).toContain('"--continue"')
    expect(resumeArgv(['codex'])).toEqual(['codex', 'resume', '--last'])
    expect(go).toContain('"resume", "--last"')
  })
})

describe('resumeArgv', () => {
  it('keeps the arguments the session was launched with', () => {
    expect(resumeArgv(['claude', '--model', 'opus'])).toEqual([
      'claude',
      '--continue',
      '--model',
      'opus',
    ])
  })

  it('puts codex’s subcommand first, where codex wants it', () => {
    expect(resumeArgv(['codex', '--model', 'gpt'])).toEqual([
      'codex',
      'resume',
      '--last',
      '--model',
      'gpt',
    ])
  })

  it('reads the program by basename', () => {
    expect(resumeArgv(['/home/x/.local/bin/claude'])).toEqual([
      '/home/x/.local/bin/claude',
      '--continue',
    ])
    expect(resumeProgram(['/home/x/.local/bin/claude'])).toBe('claude')
  })

  it('leaves alone what it cannot bring back', () => {
    expect(resumeArgv([])).toBeNull()
    expect(resumeArgv(['bash', '-l'])).toBeNull()
    expect(resumeArgv(['npm', 'run', 'dev'])).toBeNull()
    expect(resumeArgv(['opencode'])).toBeNull()
  })

  it('does not ask twice for what was asked for by hand', () => {
    expect(resumeArgv(['claude', '--continue'])).toBeNull()
    expect(resumeArgv(['claude', '-c'])).toBeNull()
    expect(resumeArgv(['claude', '--resume', 'abc'])).toBeNull()
    expect(resumeArgv(['codex', 'resume', '--last'])).toBeNull()
  })

  it('does not edit the array it was given', () => {
    const launch = ['claude', '--model', 'opus']
    resumeArgv(launch)
    expect(launch).toEqual(['claude', '--model', 'opus'])
  })
})

describe('two sessions in one directory', () => {
  const rows = [
    { launchCommand: ['claude'], cwd: '/w/api' },
    { launchCommand: ['claude'], cwd: '/w/api' },
    { launchCommand: ['claude'], cwd: '/w/web' },
    { launchCommand: ['codex'], cwd: '/w/api' },
  ]

  it('cannot both resume, because both would take the same conversation', () => {
    expect(resumeIsAmbiguous(rows, 'claude', '/w/api')).toBe(true)
    expect(restoreLaunch(rows[0], rows)).toEqual({ argv: ['claude'], resumed: false })
  })

  it('leaves the one of its kind alone', () => {
    expect(resumeIsAmbiguous(rows, 'claude', '/w/web')).toBe(false)
    expect(restoreLaunch(rows[2], rows)).toEqual({
      argv: ['claude', '--continue'],
      resumed: true,
    })
  })

  it('does not confuse two different agents in one directory', () => {
    // Two transcripts, kept by two different programs in two different places.
    expect(resumeIsAmbiguous(rows, 'codex', '/w/api')).toBe(false)
    expect(restoreLaunch(rows[3], rows)).toEqual({
      argv: ['codex', 'resume', '--last'],
      resumed: true,
    })
  })
})

describe('restoreLaunch', () => {
  it('says nothing at all about a row with no recorded argv', () => {
    // The dialog has its own two sentences for a login shell, and neither of
    // them is about resuming.
    expect(restoreLaunch({ launchCommand: [], cwd: '/w' }, [])).toBeNull()
  })

  it('reports a plain command as itself, started cold', () => {
    const row = { launchCommand: ['npm', 'run', 'dev'], cwd: '/w' }
    expect(restoreLaunch(row, [row])).toEqual({ argv: ['npm', 'run', 'dev'], resumed: false })
  })
})
